cassandra/cassandra_session.go

```diff
@@ -0,0 +1,137 @@
+package cassandra
+
+import (
+	"sync"
+	"time"
+
+	"github.com/gocql/gocql"
+	log "github.com/sirupsen/logrus"
+)
+
+// Session stores a connection to Cassandra along with associated configurations
+type Session struct {
+	session                 *gocql.Session
+	cluster                 *gocql.ClusterConfig
+	shutdown                chan struct{}
+	connectionCheckTimeout  time.Duration
+	connectionCheckInterval time.Duration
+	addrs                   string
+	logPrefix               string
+	sync.RWMutex
+}
+
+// NewSession creates and returns a CassandraSession
+func NewSession(session *gocql.Session,
+	clusterConfig *gocql.ClusterConfig,
+	shutdown chan struct{},
+	timeout time.Duration,
+	interval time.Duration,
+	addrs string,
+	logPrefix string) *Session {
+	if clusterConfig == nil {
+		panic("NewSession received nil pointer for ClusterConfig")
+	}
+	if session == nil {
+		panic("NewSession received nil pointer for session")
+	}
+
+	cs := &Session{
+		session:                 session,
+		cluster:                 clusterConfig,
+		shutdown:                shutdown,
+		connectionCheckTimeout:  timeout,
+		connectionCheckInterval: interval,
+		addrs:                   addrs,
+		logPrefix:               logPrefix,
+	}
+
+	return cs
+
+}
+
+// DeadConnectionCheck will run a query using the current Cassandra session every connectionCheckInterval
+// if it cannot query Cassandra for longer than connectionCheckTimeout it will create a new session
+//
+// if you are not using a WaitGroup in the caller, just pass in nil
+func (s *Session) DeadConnectionCheck(wg *sync.WaitGroup) {
+	if wg != nil {
+		defer wg.Done()
+	}
+
+	ticker := time.NewTicker(time.Second * s.connectionCheckInterval)
+	var totaltime time.Duration
+	var err error
+	var oldSession *gocql.Session
+
+OUTER:
+	for {
+		// connection to Cassandra has been down for longer than the configured timeout
+		if totaltime >= s.connectionCheckTimeout {
+			s.Lock()
+			for {
+				select {
+				case <-s.shutdown:
+					log.Infof("%s: received shutdown, exiting deadConnectionCheck", s.logPrefix)
+					if s.session != nil && !s.session.Closed() {
+						s.session.Close()
+					}
+					// make sure we unlock the sessionLock before returning
+					s.Unlock()
+					return
+				default:
+					log.Errorf("%s: creating new session to cassandra using hosts: %v", s.logPrefix, s.addrs)
+					if s.session != nil && !s.session.Closed() && oldSession == nil {
+						oldSession = s.session
+					}
+					s.session, err = s.cluster.CreateSession()
+					if err != nil {
+						log.Errorf("%s: error while attempting to recreate cassandra session. will retry after %v: %v", s.logPrefix, s.connectionCheckInterval.String(), err)
+						time.Sleep(s.connectionCheckInterval)
+						totaltime += s.connectionCheckInterval
```

Comment:
note that `totaltime` might become quite inaccurate, especially when the above call to `s.cluster.CreateSession()` waits for a timeout. So I think we should either measure it more accurately, or not log it further down, otherwise this can be confusing when a user analyzes the log

Reply:
Ahh, yeah. I think I will just measure it more accurately